src/mlpack/methods/ann/augmented/tasks/copy_impl.hpp

```diff
@@ -0,0 +1,108 @@
+/**
+ * @file copy_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of CopyTask class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_TASKS_COPY_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_TASKS_COPY_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "copy.hpp"
+
+#include <iostream>
+
+using std::cerr;
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+namespace tasks /* Task utilities for augmented */ {
+
+CopyTask::CopyTask(int maxLength, int nRepeats) :
+    maxLength(maxLength),
+    nRepeats(nRepeats)
+{
+  // Just storing task-specific paramters.
+}
+
+template<typename ModelType>
+double CopyTask::Evaluate(ModelType& model) {
```

Comment:
Now that I see the interface, I see a couple of problems that we have to discuss:

Right now the way we train a model e.g. RNN or LogisticRegression, etc. is we define all necessary model parameters (layer, regularization constant, etc.) and call the ``Train`` function. The ``Train`` takes the ``predictor`` and ``responses`` as input followed by an optional optimizer instance where we define the number of iterations/epochs, the threshold to terminate the optimization before we reach the maximum number of iterations, etc..


```c++
arma::mat predictor, responses;
Model model(...);

Optimizer optimizer(...);

model.Train(predictor, responses, optimizer);
```

The Evaluate function somehow replicates the functionality of the optimizer like the number of epochs. I'm not sure we should replicate the behavior since the way we train a model might be different from class to class. Like none of the existing classes can handle ``arma::field`` as input. There is no option to specify optimizer parameter, or what if we like to use the upcoming cross-validation feature as part of the evaluation pipeline. So I think instead of providing an evaluation function that takes a model and trains it we should see the Tasks as a generator that generates input/output samples for our model to train on. This enables us to train the model independently from the task. I guess besides the code replication another problem is I can't think of an easy solution to incorporate curriculum learning which might be necessary to get stable results into the class without providing a bunch of parameters for each task. So I propose the following:


```c++
CopyTask
{
 public:
  CopyTask(...) {}

  void Sample(arma::mat& predictor, arma::mat& responses)
  {
    // Generate predictor responses pair or pairs with the given parameter.
  }
}

CopyTask copyTask;
copyTask.Sample(predictor, responses);

model.Train(predictor, responses, optimizer);

```

Let me know what you think.